File: scope104/src/main/java/com/petro/scope104/presentation/list/WorkerListAdapter.java
package com.petro.scope104.presentation.list;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.petro.scope104.R;
import com.petro.scope104.databinding.ItemWorkerBinding;
import com.petro.scope104.databinding.ItemWorkerSquareBinding;
import com.petro.scope104.presentation.WorkerUi;
import com.petro.scope104.util.DateFormatHelper;

import java.util.Arrays;
import java.util.List;

public class WorkerListAdapter extends ListAdapter<WorkerUi, WorkerListAdapter.AbstractViewHolder> {
    public static final int VIEW_TYPE_NORMAL = 1;
    public static final int VIEW_TYPE_SQUARE = 2;

    private OnClickListener callback;
    private final int currentViewType;
    public WorkerListAdapter(int currentViewType) {
        super(new UserDiffCallBack());
        this.currentViewType = currentViewType;
    }

    @Override
    public int getItemViewType(int position) {
        return currentViewType;
    }

    public void setOnClickListener(OnClickListener callback) {
        this.callback = callback;
    }

    @NonNull
    @Override
    public AbstractViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (viewType){
            case VIEW_TYPE_NORMAL:
                return new WorkerViewHolder(parent, callback);
            case VIEW_TYPE_SQUARE:
                return new WorkerGridViewHolder(parent, callback);
        }
        throw new IllegalStateException();
    }


    @Override
    public void onBindViewHolder(@NonNull AbstractViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    interface OnClickListener {
        void onClick(WorkerUi clickedItem, List<View> transitionView);
    }

    private static class UserDiffCallBack extends DiffUtil.ItemCallback<WorkerUi> {

        @Override
        public boolean areItemsTheSame(@NonNull WorkerUi oldItem, @NonNull WorkerUi newItem) {
            return oldItem.getUsername().equals(newItem.getUsername());
        }

        @Override
        public boolean areContentsTheSame(@NonNull WorkerUi oldItem, @NonNull WorkerUi newItem) {
            return oldItem.equals(newItem);
        }
    }

    abstract static class AbstractViewHolder extends RecyclerView.ViewHolder{

        public AbstractViewHolder(@NonNull View itemView) {
            super(itemView);
        }
        public abstract void bind(WorkerUi item);
    }

    static class WorkerViewHolder extends AbstractViewHolder {
        private final OnClickListener onClickListener;
        private final ItemWorkerBinding binding;

        public WorkerViewHolder(@NonNull ItemWorkerBinding binding, OnClickListener onClickListener) {
            super(binding.getRoot());
            this.binding = binding;
            this.onClickListener = onClickListener;
        }

        public WorkerViewHolder(@NonNull ViewGroup parent, OnClickListener onClickListener) {
            this(ItemWorkerBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false), onClickListener);
        }

        public void bind(WorkerUi item) {
            Glide.with(binding.ivAVATAR).load(item.getAvatarUrl()).centerCrop().into(binding.ivAVATAR);
            binding.name.setText(item.getName());
            binding.date.setText(DateFormatHelper.formatDob(item.getDob()));
            binding.city.setText(item.getCity());
            binding.ivAVATAR.setTransitionName(binding.ivAVATAR.getContext().getString(R.string.avatarTransition, item.getUsername()));
            binding.name.setTransitionName(binding.name.getContext().getString(R.string.nameTransition, item.getUsername()));
            binding.root.setOnClickListener(v -> {
                if (onClickListener != null)
                    onClickListener.onClick(item, Arrays.asList(binding.ivAVATAR, binding.name));
            });
        }
    }

    static class WorkerGridViewHolder extends AbstractViewHolder {
        private final OnClickListener onClickListener;
        private final ItemWorkerSquareBinding binding;

        public WorkerGridViewHolder(@NonNull ItemWorkerSquareBinding binding, OnClickListener onClickListener) {
            super(binding.getRoot());
            this.binding = binding;
            this.onClickListener = onClickListener;
        }

        public WorkerGridViewHolder(@NonNull ViewGroup parent, OnClickListener onClickListener) {
            this(ItemWorkerSquareBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false), onClickListener);
        }

        public void bind(WorkerUi item) {
            Glide.with(binding.ivAVATAR).load(item.getAvatarUrlXXL()).centerCrop().into(binding.ivAVATAR);
            binding.name.setText(item.getName());
            binding.ivAVATAR.setTransitionName(binding.ivAVATAR.getContext().getString(R.string.avatarTransition, item.getUsername()));
            binding.name.setTransitionName(binding.name.getContext().getString(R.string.nameTransition, item.getUsername()));
            binding.root.setOnClickListener(v -> {
                if (onClickListener != null)
                    onClickListener.onClick(item, Arrays.asList(binding.ivAVATAR, binding.name));
            });
        }
    }

}

File: scope104/src/main/kotlin/com/petro/scope104/presentation/list/WorkerListAdapter.kt
package com.petro.scope104.presentation.list

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.petro.scope104.R
import com.petro.scope104.databinding.ItemWorkerBinding
import com.petro.scope104.databinding.ItemWorkerSquareBinding
import com.petro.scope104.domain.entity.WorkerEntity
import com.petro.scope104.presentation.list.WorkerListAdapter.AbstractViewHolder
import com.petro.scope104.util.DateFormatHelper
import java.util.*

class WorkerListAdapter(private val currentViewType: Int) :
    ListAdapter<WorkerEntity, AbstractViewHolder>(UserDiffCallBack()) {
    private var callback: OnClickListener? = null
    override fun getItemViewType(position: Int): Int {
        return currentViewType
    }

    fun setOnClickListener(callback: OnClickListener) {
        this.callback = callback
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AbstractViewHolder {
        when (viewType) {
            VIEW_TYPE_NORMAL -> return WorkerViewHolder(parent, callback)
            VIEW_TYPE_SQUARE -> return WorkerGridViewHolder(parent, callback)
        }
        throw IllegalStateException()
    }

    override fun onBindViewHolder(holder: AbstractViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    interface OnClickListener {
        fun onClick(clickedItem: WorkerEntity, transitionView: List<View>)
    }

    private class UserDiffCallBack : DiffUtil.ItemCallback<WorkerEntity>() {
        override fun areItemsTheSame(oldItem: WorkerEntity, newItem: WorkerEntity): Boolean {
            return oldItem.username == newItem.username
        }

        override fun areContentsTheSame(oldItem: WorkerEntity, newItem: WorkerEntity): Boolean {
            return oldItem == newItem
        }
    }

    abstract class AbstractViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        abstract fun bind(item: WorkerEntity)
    }

    internal class WorkerViewHolder(
        private val binding: ItemWorkerBinding,
        private val onClickListener: OnClickListener?
    ) : AbstractViewHolder(
        binding.getRoot()
    ) {
        constructor(parent: ViewGroup, onClickListener: OnClickListener?) : this(
            ItemWorkerBinding.inflate(
                LayoutInflater.from(parent.context), parent, false
            ), onClickListener
        )

        override fun bind(item: WorkerEntity) {
            Glide.with(binding.ivAVATAR).load(item.avatarUrl).centerCrop()
                .into(binding.ivAVATAR)
            binding.name.text = item.name
            binding.date.text = DateFormatHelper.formatDob(item.dob)
            binding.city.text = item.city
            binding.ivAVATAR.transitionName =
                binding.ivAVATAR.context.getString(R.string.avatarTransition, item.username)
            binding.name.transitionName =
                binding.name.context.getString(R.string.nameTransition, item.username)
            binding.root.setOnClickListener {
                onClickListener?.onClick(
                    item, listOf(
                        binding.ivAVATAR, binding.name
                    )
                )
            }
        }
    }

    internal class WorkerGridViewHolder(
        private val binding: ItemWorkerSquareBinding,
        private val onClickListener: OnClickListener?
    ) : AbstractViewHolder(
        binding.getRoot()
    ) {
        constructor(parent: ViewGroup, onClickListener: OnClickListener?) : this(
            ItemWorkerSquareBinding.inflate(LayoutInflater.from(parent.context), parent, false),
            onClickListener
        )

        override fun bind(item: WorkerEntity) {
            Glide.with(binding.ivAVATAR).load(item.avatarUrlXXL).centerCrop()
                .into(binding.ivAVATAR)
            binding.name.text = item.name
            binding.ivAVATAR.transitionName =
                binding.ivAVATAR.context.getString(R.string.avatarTransition, item.username)
            binding.name.transitionName =
                binding.name.context.getString(R.string.nameTransition, item.username)
            binding.root.setOnClickListener { v: View? ->
                onClickListener?.onClick(
                    item, Arrays.asList(
                        binding.ivAVATAR, binding.name
                    )
                )
            }
        }
    }

    companion object {
        const val VIEW_TYPE_NORMAL = 1
        const val VIEW_TYPE_SQUARE = 2
    }
}